File: Nim/Sources/Day01.nim
import strutils
import strformat
import sequtils
import AoCLib

const InputData = "Day01.txt"

# The state object is a hangover from VBA where it is seen as good practise
# to encapsulate Module level variables in a User Defined Type.
#
# In VBA I typically use 3 UDF, 
# - State(s) for internal state of classes, 
# - Properties(p) for class properties (acceesed using let/get)
# - Readonly(r) 'values' I want to make available from within a class as a 'constant'

type
  State = object
    Integers: seq[int]

let myPath = RawDataPath2021 & InputData
var s = new State
s.Integers= toseq(myPath.lines).map(parseInt) 
    
proc Part01() =
    
    var myResult:int=0
    for myIndex in 1..s.Integers.high:  
        if s.Integers[myindex] > s.Integers[myindex-1]:
            myresult+=1

    echo fmt"The answer to Day 01 part 1 is 1711 .  Found is {myResult}"


proc Part02() =
   
    var myResult:int=0
    var mySum:seq[int]

    for myindex in 2..s.Integers.high:
        mysum.add  s.Integers[myIndex] + s.Integers[myIndex-1]+s.Integers[myIndex-2]

    for myindex in 1..<mySum.len():
        if mysum[myIndex]>mysum[myindex-1]:
            myResult+=1

    echo fmt"The answer to Day 01 part 2 is 1743.  Found is {myResult}"        


proc Execute*()=
    Part01()
    Part02()

File: Typescript/Sources/Day01.ts
import { GetFileByLines } from "./common";
import *  as c from "./Constants";


    const InputData= "Day01.txt";

    interface State
    {
        //List<long> integers = new List<long>();

        _integers:number[]
        
        get integers() {
            return this._integers;
        }
        set length(value) {
            this._integers = value;
        }
    }

        
    


    export function Execute(): void
    {

        Part01();
        Part02();

    }

    

export function Part01() {
    var myIntegers: number[] = GetFileByLines(c.RawDataPath + "Day01RawData.txt").map(x => parseInt(x));
    var sums = GetTwoItemsThatSumToTarget(2020, myIntegers)
    var myResult = sums[0] * sums[1];
    console.log(`The answer to Day 01 Part 1 is 982464. Found is ${myResult}`);
}


    function Part01()
    {
        const fs = require('fs');
        s.integers = new List<long>(File.ReadAllLines(Constants.AoC2021Data + InputData).ToList<string>().Select(i => long.Parse(i)));
        
        myResult:number=0;
        int myIndex = 2;
        for (  myIndex = 1; myIndex< s.integers.Count;myIndex++)
        {
            if (s.integers[myIndex] > s.integers[myIndex - 1])
            {
                myResult += 1;
            }
        }
        console.log("The answer to Day 01 part 1 is 1711 .  Found is {0}", myResult);

    }


    function Part02()
    {
        s.integers = new List<long>(File.ReadAllLines(Constants.AoC2021Data + InputData).ToList<string>().Select(i => long.Parse(i)));
        List<long> mySums = new List<long>();
        
            // Compile a list of sums of 3 consecutive values
            int myIndex=0;
            for ( myIndex = 2; myIndex<s.integers.Count; myIndex++)
            {
                mySums.Add(s.integers[myIndex] + s.integers[myIndex - 1] + s.integers[myIndex - 2]);
            }
    
        // Now repeat the compare that we did for ( part 1 using mysums
        long myResult = 0;
    
        for ( myIndex = 1; myIndex < mySums.Count;myIndex++)
        {
            if (mySums[myIndex] > mySums[myIndex - 1])
            { 
                myResult += 1; 
            }
        }

        console.log("The answer to Day 01 part 2 is 1743 .  Found is {0}", myResult);

    }
}